### DISCUSSION TITLE
ESM loader not working with transpileOnly flag.

### DISCUSSION THREAD

**Discussion by thetutlage** [TypeStrong/ts-node#1721] [General]
Using the ESM loader without `transpileOnly` works fine, but breaks when using `transpileOnly`. Here's the error.

```sh
TSError: ⨯ Unable to compile TypeScript:
error TS4124: Compiler option 'module' of value 'nodenext' is unstable. Use nightly TypeScript to silence this error. Try updating with 'npm install -D typescript@next'.
```

I understand that the error comes from TypeScript asking to install the `@next` version. However, I am little confused why disabling `transpileOnly` makes it work? Is there something I am not aware of?

- `ts-node: ^10.7.0`
- `typescript: ^4.6.3`

Following is the `tsconfig.json` file.

```json
{
  "compilerOptions": {
    "target": "ES2020",
    "module": "NodeNext",
    "strict": true,
    "outDir": "./build",
    "esModuleInterop": true,
    "skipLibCheck": true,
    "noImplicitAny": false,
    "forceConsistentCasingInFileNames": true
  },
  "include": ["./**/*"],
  "exclude": ["./node_modules", "./build"],
  "ts-node": {
    "transpileOnly": true
  }
}
```

You can reproduce the issue by cloning this repo. https://github.com/thetutlage/ts-node-esm-transpile-only

- Clone the repo
- Run `npm install`
- And then run `npm start`

**Comment by cspotcode** [comment]
I'm guessing that `tsc` always raises that error when it is used with your configuration.  Note that using a different tsconfig, or using a different version of typescript, may obviously change things.  Those changes would not be caused by ts-node.

You're probably hitting an issue where ts-node is somehow omitting that error in one code path and not another.  This might be a ts-node quirk, but I believe the key takeaway is this: your TypeScript configuration raises this error, and any time the error is *not* there, you're simply getting lucky.  This behavior should not be depended upon.  You are getting that error for a reason, and you should read what it says and change your configuration accordingly.

If you want a way to suppress individual diagnostic codes in ts-node, we do have an option for that: `ignoreDiagnostics`
https://typestrong.org/ts-node/docs/options#ignorediagnostics

**Reply by thetutlage** [reply]
> You are getting that error for a reason, and you should read what it says and change your configuration accordingly

I did read it :) I think you missed the point of why I created the thread

> time the error is not there, you're simply getting lucky

🤷🏻‍♂️. Heard for the first time that we can even try our luck with compilers.

Anyways, I know that TypeScript gives the error, but `ts-node` doesn't, was not something I was expecting

**Reply by cspotcode** [reply]
Ok, you are wanting ts-node to *always* raise that error, and it might be a bug that ts-node is somehow suppressing it.

I haven't looked at the relevant code in a while, but when typechecking, we only throw errors for diagnostics that happen within the file being `require()`d or `import`ed.  #853  Perhaps that is related?

**Reply by thetutlage** [reply]
Yup, exactly, that is what I was looking for :)
